pkg/coscheduling/core/core.go

```diff
@@ -0,0 +1,241 @@
+	snapshotSharedLister framework.SharedLister
+	// scheduleTimeout is the default time when group scheduling.
+	// If podgroup's ScheduleTimeoutSeconds set, that would be used.
+	scheduleTimeout *time.Duration
+	// lastDeniedPG store the pg name if a pod can not pass pre-filer,
+	// or anyone of the pod timeout
+	lastDeniedPG *gochache.Cache
+	// pgLister is podgroup lister
+	pgLister pglister.PodGroupLister
+	// podLister is pod lister
+	podLister listerv1.PodLister
+	// reserveResourcePercentage is the reserved resource for the max finished group, range (0,100]
+	reserveResourcePercentage int32
+	sync.RWMutex
+}
+
+// NewPodGroupManager create a new operation object
+func NewPodGroupManager(pgClient pgclientset.Interface, snapshotSharedLister framework.SharedLister, scheduleTimeout *time.Duration,
+	pgInformer pginformer.PodGroupInformer, podInformer informerv1.PodInformer) *PodGroupManager {
+	pgMgr := &PodGroupManager{
+		pgClient:             pgClient,
+		snapshotSharedLister: snapshotSharedLister,
+		scheduleTimeout:      scheduleTimeout,
+		pgLister:             pgInformer.Lister(),
+		podLister:            podInformer.Lister(),
+		lastDeniedPG:         gochache.New(3*time.Second, 3*time.Second),
+	}
+	return pgMgr
+}
+
+// PreFilter pre-filter pod include:
+// 1. cluster resource
+// 2. last denied pod to increase schedule qps
+// 3. check resource for max finished group, evict others if it can not run
+// 4. check pod priority, if it can not run, evict other groups
+func (pgMgr *PodGroupManager) PreFilter(ctx context.Context, pod *corev1.Pod) error {
+	klog.V(10).Infof("Pre-filter %v", pod.Name)
+	pgFullName, pg := pgMgr.GetPodGroup(pod)
+	if pg == nil {
+		return nil
+	}
+	if _, ok := pgMgr.lastDeniedPG.Get(pgFullName); ok {
+		err := fmt.Errorf("pod with pgName: %v last failed in 20s, deny", pgFullName)
+		klog.V(6).Info(err)
+		return err
+	}
+	pods, err := pgMgr.podLister.Pods(pod.Namespace).List(
+		labels.SelectorFromSet(labels.Set{util.PodGroupLabel: util.GetPodGroupLabel(pod)}),
+	)
+	if err != nil {
+		return fmt.Errorf("podLister list pods failed: %v", err)
+	}
+	if len(pods) < int(pg.Spec.MinMember) {
+		return fmt.Errorf("cannot found engough pods, "+
+			"current pods number: %v, minMember of group: %v", len(pods), pg.Spec.MinMember)
+	}
+	return nil
+}
+
+// Permit permits a pod to run, if the minMember match, it would send a signal to chan.
+func (pgMgr *PodGroupManager) Permit(ctx context.Context, pod *corev1.Pod, nodeName string) (bool, error) {
+	pgFullName, pg := pgMgr.GetPodGroup(pod)
+	if pgFullName == "" {
+		return true, util.ErrorNotMatched
+	}
+	if pg == nil {
+		// once we have admission check, a Pod targeting non-exisitng PodGroup won't be created.
+		// So here it just serves as a sanity check.
+		return false, fmt.Errorf("PodGroup not found")
+	}
+	if pg.Status.Phase == "" || pg.Status.Phase == v1alpha1.PodGroupPending {
+		pg.Status.Phase = v1alpha1.PodGroupPreScheduling
```

Comment:
Where need to use `v1alpha1.PodGroupPreScheduling`? This state is not synchronized to the apiserver.

Reply:
Thanks, done

Reply:
I am not sure if we really need to provide the state of PreScheduling. 
em.. I think we shouldn't  sync the state to the apiserver in scheduling cycle because the network access time.

Reply:
Yes. This is also my concern though not every pod would reach here. WDYT if open a goroutine to do this?

Reply:
I think the states of `Scheduling` and `Scheduled` is enough for us.
The info of pod added in waiting map can be find in log if we need it to debug. 
Other concern is that `PreScheduling` may be mistaken for waiting in the queue (like pending) if they don't read the code.

Reply:
IMO Prescheduling is the status when number of pod is enough and start scheduling the group

Reply:
> IMO Prescheduling is the status when number of pod is enough and start scheduling the group

That sounds to me it means "at least one of the Pods" has started the scheduling cycle?

My concern is that we obtain PG object from PGLister, so it's possible that: when the 1st pod of a Pod Group marks the PG status to be `PreScheduling`, the status is not necessarily reflected into the PGLister, so the 2nd (or even 3rd, ...) pod of that PodGroup may still read the status as `Pending`, and hence re-trigger the update operation - which is an overhead to API server.

So if this field is not that useful to end-user, we may make a tradeoff here to eliminate it. (I assume we don't have logic relying on this field)

Reply:
@Huang-Wei @denkensk 

Thanks. Have removed .

I think maybe updating group status in podgroup controller is better to avoid blocking the schedule cycle, would add the logic to controller in another PR


PTAL